### ISSUE TITLE
Add function to distinguish between segmented blobs and stream blobs to RDB$BLOB_UTIL

### ISSUE THREAD

**Issue by mrotteveel** [FirebirdSQL/firebird#7572]
Given `RDB$BLOB_UTIL.SEEK()` only works on stream blobs, it would be handy to have a function to distinguish between segmented and stream blobs. For example `RDB$BLOB_UTIL.IS_STREAM(BLOB)` which returns `TRUE` if it is a stream blob, or `FALSE` if it is a segmented blob.

**Comment by mrotteveel** [comment]
As an aside, implementing #4711 would reduce the need for this.

**Comment by AlexPeshkoff** [comment]
Segmented vs. stream differ only in a way how blob was opened. I.e. caller already has full and precise information how is blob accessed.

**Comment by mrotteveel** [comment]
@AlexPeshkoff That is not correct. The blob type is stored together with the blob, as segmented blobs are stored differently than streamed blobs, and a blob is opened as that original type. Attempts to seek in a blob that was originally created as segmented will fail. The code (and the author of the code) using `RDB$BLOB_UTIL.SEEK` is not necessarily in control of how blobs are created, or were created in the past, so there needs to be a way to determine the blob type so the code can use alternative handling (e.g. using `RDB$BLOB_UTIL.READ_DATA` to move forward), or some other form of fallback.

Add to that that things like creating a blob by (implicit or explicit) casting will also always create a segmented blob, and you get situations were stored data, or data passed in as parameter to a stored procedure may not always be a stream blob.

A simple example in Java:

```java
public class BlobUtilSeek {

    private static final boolean USE_STREAM_BLOB = false;

    public static void main(String[] args) throws SQLException {
        setup();
        createBlob();

        seekInBlob();
    }

    private static void setup() throws SQLException {
        try (var connection = DefaultDb.createDefaultConnection(DefaultDb.JDBC_URL_50);
             var statement = connection.createStatement()) {
            statement.execute("""
                    recreate table blob_test (
                      id integer generated always as identity primary key,
                      val blob sub_type binary
                    )
                    """);
        }
    }

    private static void createBlob() throws SQLException {
        var props = new Properties();
        props.setProperty("useStreamBlobs", String.valueOf(USE_STREAM_BLOB));
        try (var connection = DefaultDb.createDefaultConnection(DefaultDb.JDBC_URL_50, props);
             var pstmt = connection.prepareStatement("insert into blob_test (val) values (?)")) {
            var data = new byte[64 * 1024];
            ThreadLocalRandom.current().nextBytes(data);
            pstmt.setBytes(1, data);
            pstmt.execute();
        }
    }

    private static void seekInBlob() throws SQLException {
        try (var connection = DefaultDb.createDefaultConnection(DefaultDb.JDBC_URL_50);
             var statement = connection.createStatement();
             var rs = statement.executeQuery("""
                     execute block returns (fragment varbinary(512))
                     as
                     declare val type of column blob_test.val;
                     declare handle integer;
                     begin
                       for select val from blob_test into val do
                       begin
                         handle = rdb$blob_util.open_blob(val);
                         rdb$blob_util.seek(handle, 0, 32765);
                         fragment = rdb$blob_util.read_data(handle, 512);
                         execute procedure rdb$blob_util.close_handle(handle);
                         suspend;
                       end
                     end
                     """)) {
            while (rs.next()) {
                System.out.printf("Data length: %d", rs.getBytes(1).length);
            }
        }
    }
}
```

Running `BlobUtilSeek` will fail with `USE_STREAM_BLOB = false` (which *only* determines how blobs are created by Jaybird), and will succeed when `USE_STREAM_BLOB = true`.

For reference the class `DefaultDb` is:

```java
public class DefaultDb {
    public static final String JDBC_URL_30 = "jdbc:firebirdsql://localhost/E:/db/fb3/fb3testdatabase.fdb";
    public static final String JDBC_URL = JDBC_URL_30;
    public static final String JDBC_URL_25 = "jdbc:firebirdsql://localhost/E:/db/testdatabase.fdb?encoding=NONE";
    public static final String JDBC_URL_40 = "jdbc:firebirdsql://localhost/E:/db/fb4/fb4testdatabase.fdb";
    public static final String JDBC_URL_50 = "jdbc:firebirdsql://localhost/E:/db/fb5/fb5testdatabase.fdb";
    public static final String DS_URL = "//localhost/E:/db/testdatabase.fdb";
    public static final String USER_NAME = "SYSDBA";
    public static final String PASSWORD = "masterkey";

    public static Connection createDefaultConnection() throws SQLException {
        return createDefaultConnection(JDBC_URL);
    }

    public static Connection createDefaultConnection(String url) throws SQLException {
        return DriverManager.getConnection(url, USER_NAME, PASSWORD);
    }

    public static Connection createDefaultConnection(Properties props) throws SQLException {
        return createDefaultConnection(JDBC_URL, props);
    }

    public static Connection createDefaultConnection(String url, Properties props) throws SQLException {
        Properties finalProps = new Properties();
        finalProps.putAll(props);
        finalProps.setProperty("user", USER_NAME);
        finalProps.setProperty("password", PASSWORD);
        return DriverManager.getConnection(url, finalProps);
    }
}
```

**Comment by mrotteveel** [comment]
Reproduction in SQL:

```sql
recreate table blob_test (
  id integer generated always as identity primary key,
  val blob sub_type binary
);

insert into blob_test (val) values (cast('abcdefghijklmnopqrstuvwxyz' as blob sub_type binary));

set term #;
execute block returns (fragment varbinary(512))
as
declare val type of column blob_test.val;
declare handle integer;
begin
  for select val from blob_test into val do
  begin
    handle = rdb$blob_util.open_blob(val);
    rdb$blob_util.seek(handle, 0, 1);
    fragment = rdb$blob_util.read_data(handle, 5);
    execute procedure rdb$blob_util.close_handle(handle);
    suspend;
  end
end#
set term ;#
```

Result:

> Statement failed, SQLSTATE = 42000
invalid BLOB type for operation
-At function 'RDB$BLOB_UTIL.SEEK'
At block line: 9, col: 5

**Comment by AlexPeshkoff** [comment]
On 5/10/23 10:37, Mark Rotteveel wrote:
>
> Reproduction in SQL:
>
> recreate table blob_test (
>    idinteger  generated alwaysas  identityprimary key,
>    val blob sub_type binary
> );
>
> insert into  blob_test (val)values  (cast('abcdefghijklmnopqrstuvwxyz'  as  blob sub_type binary));
>
> set  term#;
> execute block returns (fragment varbinary(512))
> as
> declare val type of columnblob_test.val;
> declare handleinteger;
> begin
>    forselect  valfrom  blob_test into val do
>    begin
>      handle=  rdb$blob_util.open_blob(val);
>      rdb$blob_util.seek(handle,0,1);
>      fragment=  rdb$blob_util.read_data(handle,5);
>      execute procedure rdb$blob_util.close_handle(handle);
>      suspend;
>    end
> end#
> set  term ;#
>
> Result:
>
>     Statement failed, SQLSTATE = 42000
>     invalid BLOB type for operation
>     -At function 'RDB$BLOB_UTIL.SEEK'
>     At block line: 9, col: 5
>

Mark sorry but this reproduces nothing. Here is how engine decides 
access blob as segmented or streamed. Pay attention - nothing except BPB 
is taken into an account.

     const SSHORT type = gds__parse_bpb2(bpb_length,
                            bpb,
                            &from,
                            &to,
reinterpret_cast<USHORT*>(&from_charset),    // safe - alignment not changed
reinterpret_cast<USHORT*>(&to_charset),      // safe - alignment not changed
                            NULL, NULL, NULL, NULL);

     blb* blob = allocate_blob(tdbb, transaction);

     if (type & isc_bpb_type_stream)
         blob->blb_flags |= BLB_stream;

**Comment by AlexPeshkoff** [comment]
My excuses - yes, format of blob really stored in database.

**Comment by mrotteveel** [comment]
> My excuses - yes, format of blob really stored in database.

👍

**Comment by aafemt** [comment]
Isn't it simplier to implement seek() on segmented BLOBs?..

**Comment by mrotteveel** [comment]
> Isn't it simplier to implement seek() on segmented BLOBs?..

Maybe, although seeking from end, or a backwards seek relative to the current position is problematic because of the segments (i.e. you need to know where the segment sizes are to avoid interpreting them as data). A forward seek is probably doable, both from start and relative to current position.

As an aside, I don't understand why we don't ditch segmented blobs, and store everything as stream blobs. I don't see the benefits of segmented blobs.

**Comment by aafemt** [comment]
Segmented BLOBs has benefit that segments are retrieved exactly as they were stored. Handling of streams may be complex if data is naturally segmented (UTF-8 text for example).

**Comment by mrotteveel** [comment]
> Segmented BLOBs has benefit that segments are retrieved exactly as they were stored. Handling of streams may be complex if data is naturally segmented (UTF-8 text for example).

As far as I recall, if a segmented blob is stored in 5 segments of 512 bytes, and you ask for a segment of 32KB, you receive one segment of 2560, and if you ask for a segment of 100 bytes, you receive a segment of 100 bytes, no matter how it was stored.

**Comment by aafemt** [comment]
It is not true. You'll get 512 bytes (or size of buffer).

**Comment by aafemt** [comment]
> That is not what I see in get_segment in server.cpp

This is what I see in BLB_get_segment() in blb.cpp. Server is a special case: it send as much segments as it can to client where they are cached.

> If MBCS data do not work correctly in stream blobs

MBCS (AFAIK) works correctly in server. I'm talking about client applications which may be not ready to handle thorn characters from streamed BLOB.

**Comment by mrotteveel** [comment]
> > That is not what I see in get_segment in server.cpp
> 
> This is what I see in BLB_get_segment() in blb.cpp. Server is a special case: it send as much segments as it can to client where they are cached.

It sends the **data** as a single byte array, not as a collection of segments, irrespective of how many segments (or partial segments) it reads originally, the only "oddity" it seems to have that occasionally it sends 2 bytes less than requested.

> > If MBCS data do not work correctly in stream blobs
> 
> MBCS (AFAIK) works correctly in server. I'm talking about client applications which may be not ready to handle thorn characters from streamed BLOB.

Given actual segment sizes are ignored for the data returned to the client (at least over network connections), but data of the requested size (or requested size - 2, or the remaining bytes) is sent, they already need to handle this (and besides, that is really a non issue, as any competently implemented character set decoder will buffer the data of incomplete characters and wait for more data to complete that character.

**Comment by aafemt** [comment]
Server does send segment length in get_segment():
```
		p[-2] = (UCHAR) length;
		p[-1] = (UCHAR) (length >> 8);
		p += length;
		buffer_length -= length;
```

**Comment by mrotteveel** [comment]
You're right, I missed that in my own implementation, but that is not necessarily the actual segment size, just the remaining size of the segment to fit in the requested length. For example, if create a blob with segments of 255 bytes, and read per 100, the first 3 response I receive are):

- 2x 100 bytes, with length 98 and 98 bytes data, 
- 10 bytes with length 59, 59 bytes of data, length 37, 37 bytes of data, for total 96 bytes of data.

In other words, the size of stored segments has little bearing on what you get returned, though the implementation is slightly inefficient when it crosses the boundary of a segment, because of the additional length information (I guess it would be possible to optimize this by coalescing segments)

**Comment by mrotteveel** [comment]
But although interesting, this is not really that relevant for this issue itself ;)

**Comment by aafemt** [comment]
It is internal machinery between server and client. Client application either gets whole segment if buffer is big enough or piece of it otherwise.

**Comment by sim1984** [comment]
> Isn't it simplier to implement seek() on segmented BLOBs?..

Even if you implement a search on segmented blobs, this does not cancel the desire to know which particular blob is open. I might just be wondering how my applications save blobs, which was written before me.
